csrc/dynamic_transform.cpp

```diff
@@ -376,13 +380,37 @@ void DynamicTransformConcretizer::concretize() {
   // Set output IterTypes for dynamic resize ops
   concretizeResize();
 
-  // Finally, propagate concretized domains
+  // The methods above do not traverse the graph. Instead they fill in
+  // root->rfactor expressions by replacing the dynamic reshaped TV with a
+  // static reshaped one, and by registering concretization of dynamic Resized
+  // IterDomains. From this point forward, we will not modify any TensorView
+  // expressions. This restriction makes it safe for us to to traverse forward
+  // through the graph and mutate IterDomains and TensorDomains in order to
+  // properly propagate IterTypes and concretized extent expressions, without
+  // breaking the topological ordering of these expressions.
+  //
+  // When propagating IterTypes across expressions, we need to know the producer
+  // IterDomains corresponding to a consumer ID. This mapping helps facilitate
+  // this and is used later in mutate(IterDomain*).
   auto all_stmts = StmtSort::getStmts(info_->fusion(), true);
-  for (auto stmt : all_stmts) {
-    if (stmt->isA<Val>()) {
-      mutate(stmt);
```

Comment:
Instead of only mutating `Val`s, we now mutate `Expr`s as well, which replaces the `Expr` in place if any inputs or outputs have changed. Note that outputs of `Expr`s are typically registered for mutation after their definition has been mutated, so we should be careful updating a `Val` that has a definition. But of course we should be careful in that case in the existing code too.